### PR TITLE
support connection to a GraphDatabaseService instance

### PR DESCRIPTION
**Pull request by YoannBuch**
See #141

### CODE REVIEW

File: liquigraph-core/src/main/java/org/liquigraph/core/configuration/ConfigurationBuilder.java [hunk 1]
```diff
@@ -78,8 +81,9 @@ public ConfigurationBuilder withUri(String uri) {
     }
 
     /**
-     * Specifies the data source of the graph database instance.
-     * Alternatively, you can set the URI
+     * Specifies the {@code GraphDataseService} instance.
```
Comment (fbiville) [inline liquigraph-core/src/main/java/org/liquigraph/core/configuration/ConfigurationBuilder.java]: typo: GraphDataseService

File: liquigraph-core/src/main/java/org/liquigraph/core/configuration/ConfigurationBuilder.java [hunk 2]
```diff
@@ -78,8 +81,9 @@ public ConfigurationBuilder withUri(String uri) {
...
+     * Alternatively, you can set the URI or a
+     * {@code GraphDataseService}.
```
Comment (fbiville) [inline liquigraph-core/src/main/java/org/liquigraph/core/configuration/ConfigurationBuilder.java]: typo: GraphDataseService

File: liquigraph-core/src/main/java/org/liquigraph/core/configuration/ConfigurationBuilder.java (hunk 1 above)
Comment (fbiville) [inline liquigraph-core/src/main/java/org/liquigraph/core/configuration/ConfigurationBuilder.java]: Shouldn't it be DataSource (and not GDS) here?

File: liquigraph-core/src/main/java/org/liquigraph/core/configuration/ConfigurationBuilder.java [hunk 3]
```diff
@@ -90,6 +94,18 @@ public ConfigurationBuilder withDataSource(DataSource dataSource) {
     }
 
     /**
+     * Specifies the data source of the graph database instance.
```
Comment (fbiville) [inline liquigraph-core/src/main/java/org/liquigraph/core/configuration/ConfigurationBuilder.java]: s/the data source of // ?

File: liquigraph-core/src/main/java/org/liquigraph/core/configuration/ConnectionConfigurationByGraphDatabaseservice.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.UUID;
+
+import static com.google.common.base.Throwables.propagate;
+
+class ConnectionConfigurationByGraphDatabaseservice implements ConnectionConfiguration {
+
+    private final String uri;
+    private final Properties driverProperties;
+
+    public ConnectionConfigurationByGraphDatabaseservice(GraphDatabaseService graphDatabaseService) {
```
Comment (fbiville) [inline liquigraph-core/src/main/java/org/liquigraph/core/configuration/ConnectionConfigurationByGraphDatabaseservice.java]: minor: s of service should be in upper case

File: liquigraph-core/src/main/java/org/liquigraph/core/configuration/ConnectionConfigurationByGraphDatabaseservice.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+        // databaseName is what the driver uses to link the URI to the
```
Comment (fbiville) [inline liquigraph-core/src/main/java/org/liquigraph/core/configuration/ConnectionConfigurationByGraphDatabaseservice.java]: move this inline comment to proper documentation section on `generateUniqueDatabaseName`?

File: liquigraph-core/src/main/java/org/liquigraph/core/configuration/validators/DatasourceConfigurationValidator.java [hunk 6]
```diff
@@ -28,9 +29,20 @@
...
-    public Collection<String> validate(Optional<String> uri, Optional<DataSource> dataSource) {
-        if (uri.isPresent() == dataSource.isPresent()) {
-            return ImmutableList.of("Exactly one of JDBC URI or DataSource need to be configured");
+    public static final String INCORRECT_CONFIGURATION_OPTION_ERROR_MESSAGE = "Exactly one of JDBC URI, JDBC DataSource or GraphDatabaseService needs to be configured";
+
+    public Collection<String> validate(Optional<String> uri, Optional<DataSource> dataSource, Optional<GraphDatabaseService> database) {
+
+        int presentInstanceCount = 0;
+
+        for (Optional<?> optional : ImmutableList.of(uri, dataSource, database)) {
```
Comment (fbiville) [inline liquigraph-core/src/main/java/org/liquigraph/core/configuration/validators/DatasourceConfigurationValidator.java]: Optional: you can filter with Guava and count at the end (your version is fine as well)

File: liquigraph-core/src/test/java/org/liquigraph/core/configuration/ConfigurationBuilderTest.java [hunk 7]
```diff
@@ -57,7 +58,7 @@ public void fails_on_non_existing_changelog() {
     @Test
     public void fails_on_null_uri() {
         thrown.expect(RuntimeException.class);
-        thrown.expectMessage("- Exactly one of JDBC URI or DataSource need to be configured");
+        thrown.expectMessage("- " + INCORRECT_CONFIGURATION_OPTION_ERROR_MESSAGE);
```
Comment (fbiville) [inline liquigraph-core/src/test/java/org/liquigraph/core/configuration/ConfigurationBuilderTest.java]: optional: duplicate the message here, it makes the test code more straightforward to read (although it'll fail whenever the error change, but that's kind of an API change so it's OK imo)

File: liquigraph-core/src/test/java/org/liquigraph/core/configuration/ConfigurationBuilderTest.java [hunk 8]
```diff
@@ -85,7 +86,7 @@ public void fails_on_unsupported_protocol() {
     public void add_ups_all_misconfiguration_errors() {
         thrown.expect(RuntimeException.class);
         thrown.expectMessage(" - 'masterChangelog' should not be null");
-        thrown.expectMessage(" - Exactly one of JDBC URI or DataSource need to be configured");
+        thrown.expectMessage(" - " + INCORRECT_CONFIGURATION_OPTION_ERROR_MESSAGE);
```
Comment (fbiville) [inline liquigraph-core/src/test/java/org/liquigraph/core/configuration/ConfigurationBuilderTest.java]: same as above

File: liquigraph-core/src/test/java/org/liquigraph/core/configuration/ConnectionConfigurationByGraphDatabaseserviceTest.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+        // Extract database name from URI, expected format is
+        // "jdbc:neo4j:instance:databaseName"
+        String uri = configuration.getUri();
+        assertThat(uri)
+                .isNotNull()
+                .isNotEmpty()
+                .contains(":");
+        String[] splittedUri = uri.split(":");
+        String databaseName = splittedUri[splittedUri.length - 1];
```
Comment (fbiville) [inline liquigraph-core/src/test/java/org/liquigraph/core/configuration/ConnectionConfigurationByGraphDatabaseserviceTest.java]: optional: extract lines 45-51 to a method (a well-named extracted method is usually better than a possibly out-of-sync comment)

File: liquigraph-core/src/main/java/org/liquigraph/core/configuration/validators/DatasourceConfigurationValidator.java (hunk 6 above)
Comment (YoannBuch) [inline liquigraph-core/src/main/java/org/liquigraph/core/configuration/validators/DatasourceConfigurationValidator.java]: I never used Guava, so you'll let me know if my new changes make sense. I couldn't find a simple way to filter based on the presence of an optional instance.

**Comment by fbiville** [comment]
Merged in https://github.com/fbiville/liquigraph/commit/2146b8df823d15d72e47549a2ff4f14d3ee28fff.
We cannot merge this in `master_3.x` as long as the underlying JDBC driver does not support embedded instances. We'll need to remember to cherry-pick when it does, though.

**Pull request closed without merging**